test/framework/control_plane.go

```diff
@@ -34,213 +34,284 @@ import (
-		if err := mgmtClient.Create(ctx, input.Cluster); err != nil {
+		if err := input.Creator.Create(ctx, input.Cluster); err != nil {
 			fmt.Printf("%+v\n", err)
 			return err
 		}
 		return nil
-	}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
+	}, intervals...).Should(Succeed())
+}
 
-	By("creating related resources")
-	for i := range input.RelatedResources {
-		obj := input.RelatedResources[i]
-		By(fmt.Sprintf("creating a/an %s resource", obj.GetObjectKind().GroupVersionKind()))
-		Eventually(func() error {
-			return mgmtClient.Create(ctx, obj)
-		}, input.CreateTimeout, eventuallyInterval).Should(BeNil())
-	}
+// CreateKubeadmControlPlaneInput is the input for CreateKubeadmControlPlane.
+type CreateKubeadmControlPlaneInput struct {
+	Creator         Creator
+	ControlPlane    *controlplanev1.KubeadmControlPlane
+	MachineTemplate runtime.Object
+}
 
+// CreateKubeadmControlPlane creates the control plane object and necessary dependencies.
+func CreateKubeadmControlPlane(ctx context.Context, input CreateKubeadmControlPlaneInput, intervals ...interface{}) {
 	By("creating the machine template")
-	Expect(mgmtClient.Create(ctx, input.MachineTemplate)).To(Succeed())
+	Expect(input.Creator.Create(ctx, input.MachineTemplate)).To(Succeed())
 
 	By("creating a KubeadmControlPlane")
 	Eventually(func() error {
-		err := mgmtClient.Create(ctx, input.ControlPlane)
+		err := input.Creator.Create(ctx, input.ControlPlane)
 		if err != nil {
 			fmt.Println(err)
 		}
 		return err
-	}, input.CreateTimeout, 10*time.Second).Should(BeNil())
+	}, intervals...).Should(Succeed())
+}
+
+// CreateMachineDeploymentInput is the input for CreateMachineDeployment.
+type CreateMachineDeploymentInput struct {
+	Creator                 Creator
+	MachineDeployment       *clusterv1.MachineDeployment
+	BootstrapConfigTemplate runtime.Object
+	InfraMachineTemplate    runtime.Object
+}
+
+// CreateMachineDeployment creates the machine deployment and dependencies.
+func CreateMachineDeployment(ctx context.Context, input CreateMachineDeploymentInput) {
+	By("creating a core MachineDeployment resource")
+	Expect(input.Creator.Create(ctx, input.MachineDeployment)).To(Succeed())
+
+	By("creating a BootstrapConfigTemplate resource")
+	Expect(input.Creator.Create(ctx, input.BootstrapConfigTemplate)).To(Succeed())
+
+	By("creating an InfrastructureMachineTemplate resource")
+	Expect(input.Creator.Create(ctx, input.InfraMachineTemplate)).To(Succeed())
+}
 
+// WaitForMachineDeploymentNodesToExistInput is the input for WaitForMachineDeploymentNodesToExist.
+type WaitForMachineDeploymentNodesToExistInput struct {
+	Lister            Lister
+	Cluster           *clusterv1.Cluster
+	MachineDeployment *clusterv1.MachineDeployment
+}
+
+// WaitForMachineDeploymentNodesToExist waits until all nodes associated with a machine deployment exist.
+func WaitForMachineDeploymentNodesToExist(ctx context.Context, input WaitForMachineDeploymentNodesToExistInput, intervals ...interface{}) {
```

Comment:
Function name and comment doesn't match the behavior now. This will return after the Machine resources have been created for the related MachineDeployment, which will be much earlier than when the nodes are actually created.

Reply:
right, next is noderefs. i thought there was more. plus i need to pick out the machine set based on the machine deployment selector

Reply:
AH this surfaced a different bug with the control plane nodes as well! I think we'll have better luck with flakes in the future. Thanks for your patience while my brain struggled today.